MongoDAO

jlhughes opened the pull request:
MongoDAO moved to own repository foam-mongodb. 

The remaining changes here are for supported additional databases such as that provided by foam-mongodb. 

Also, some miscellaneous fixes and cleanup, noticed along the way.

jlhughes committed:
MongoDAO - re-implement to support all CRUD operations

jlhughes committed:
OrDAO - invert delegate setup so puts go to delegate

jlhughes committed:
pom - move mongodb java dependencies to mongodb pom so only acquired when mongo in use

jlhughes committed:
MongoDAO - additional documentation

jlhughes committed:
intro-to-flow.flow - correct markup syntax errors

jlhughes committed:
add flow doc

jlhughes committed:
remove accidentally commit temporary files

jlhughes committed:
MongoDBService.js - remove commented log calls

jlhughes committed:
MongoDAO - create Filter BsonDocument with Object rather than String value, conversion to String not necessary

jlhughes committed:
MongoDAO - add logic to support using MongoDB in a real FOAM application

jlhughes committed:
OrDAO - revert, changes not used

jlhughes commented on src/foam/dao/F3FileJournal.js:
```diff
@@ -116,9 +116,9 @@ foam.CLASS({
           pm.log(x);
           if ( getFailCount() == 0 ) {
-            getLogger().info("Replay complete", getFilename(), "processed", getPassCount(), "of", getFailCount()+getPassCount(), "in", Duration.ofMillis(pm.getTime()));
+            getLogger().info("Replay complete", getFilename(), "processed", passCount.get(), "of", failCount.get()+passCount.get(), "in", Duration.ofMillis(pm.getTime()));
```
Noticed, that many counts are not reported as the model properties are not yet set when the log is built.

jlhughes committed:
MongoDBService - update documentation about listener for cluster updates

jlhughes committed:
documenation - cluster/replicate-set

jlhughes committed:
mongodb-test - move all journals to deployment/

jlhughes committed:
flow - add macos installation notes

jlhughes committed:
flow - documentation warning that this is not a journal and to be aware of replay behaviour

lchanmann commented on src/foam/dao/mongodb/MongoDAO.flow:
```diff
@@ -0,0 +1,64 @@
+
+<h2>Cluster / Replica Sets</h2>
+<p>MongoDB supports change notification.  A listener can be
+notified of database or collection chaanges in a <b>Replica Set</b></p>
```
```suggestion
notified of database or collection changes in a <b>Replica Set</b></p>
```

jlhughes replied:
fixed.

lchanmann commented on src/foam/dao/mongodb/MongoDBService.js:
```diff
@@ -0,0 +1,130 @@
+  methods: [
+    {
+      name: 'getDatabase',
+      args: 'X x, String collectionName',
+      javaType: 'MongoDatabase',
+      javaCode: `
+      return getClient().getDatabase(getDbName());
```
`collectionName` arg is not being used.

jlhughes replied:
Added documentation

lchanmann commented on src/foam/dao/mongodb/MongoDAO.js:
```diff
@@ -0,0 +1,348 @@
+
+            FObject obj = parser_.get().parseString(cursor.next().toJson(), getOf().getObjClass());
+
+            if ( ( predicate == null ) || predicate.f(obj) ) {
+              decorated.put(obj, sub);
+            }
```
The decorated sink should already be wrapped in PredicatedSink if there is a predicate.
```suggestion
            decorated.put(obj, sub);
```

jlhughes replied:
> The decorated sink should already be wrapped in PredicatedSink if there is a predicate.

This occurs in AbstractSink.decorateSink_ called on line 219.

jlhughes committed:
MongoDAO - json output - disable short names - makes database easier to query by humans

lchanmann commented on src/foam/dao/DatabaseType.js:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
```
year 2025

jlhughes replied:
updated

lchanmann commented on src/foam/dao/mongodb/MongoDAO.js:
```diff
@@ -0,0 +1,355 @@
+    'org.bson.BsonDocumentReader',
+    'org.bson.BsonInt64',
+    'org.bson.BsonType',
+    'org.bson.Document',
```
There are some unused imports, if we can clean them of before merging.

jlhughes replied:
cleaned up.

lchanmann commented on src/foam/dao/mongodb/MongoDBService.js:
```diff
@@ -0,0 +1,130 @@
+    {
+      name: 'reload',
+      javaCode: `
+        MongoClient client = getClient();
+        initialize(getX());
+        if ( client != null ) {
+          client.close();
+        }
```
should it stop then start instead?

jlhughes replied:
I believe it should start the new to not interfere with current connection operations.  
Future logic could wait for idle before close. but in either case, acquire the new connection first.

lchanmann commented on src/foam/dao/mongodb/MongoDBService.js:
```diff
@@ -0,0 +1,130 @@
+      `,
+    },
+    {
+      name: 'stop',
+      javaCode: `
+        MongoClient client = getClient();
+        if ( client != null ) {
+          client.close();
+        }
```
MongoDBService keeps an open client connection to mongodb. Can we call `stop()` on graceful shutdown to close the connection?

Just a thought, I'm thinking of adding a shutdown hook in Boot.java to call `stop()` on instances of COREService.

jlhughes replied:
yes, shutdown hook has been on my todo for some time, and it was the reason I added 'stop' to COREService.  Will add.

jlhughes replied:
@lchanmann see shutdown hook: https://github.com/kgrgreer/foam3/pull/3740

lchanmann reviewed (commented):
Excellent!

jlhughes committed:
flow - correct typos in documenation

jlhughes committed:
DatabaseType - update header

jlhughes committed:
MongoDAO - cleanup imports

jlhughes committed:
MongoDBService - add documentation on optional collectionName on getDatabase method

jlhughes committed:
DDAO - correct mismatch in getMongoDAO calls in previous commits

lchanmann committed:
Prevent failing to start a dao on db add index command error

eg. serviceConfigDAO where the easydao config has addPropertyIndex on
the multipart id itself

lchanmann committed:
Fix json parsing for subclasses

records in mongodb collection lost subclass type on replay

lchanmann committed:
Fix duplicate replay on local and non-local dao

since CSpecFactory is looping through the inner dao stack

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into MongoDAO

jlhughes committed:
MongoDAO moved to own repo foam-mongodb

jlhughes committed:
Merge branch 'MongoDAO' of github.com:kgrgreer/foam3 into MongoDAO

jlhughes committed:
flow - correct 'use' sectione
